### PR TITLE
[example] Micro ViT

### PR DESCRIPTION
**Pull request by blefaudeux**
Co-authored-by: Sean Naren <sean@grid.ai>

## What does this PR do?
adding a really nice example from @SeanNaren, instantiate a ViT and train on CiFAR 

TODO:
- [x] try to remove some dependencies (einops maybe)
- [x] add an inference example in the end + show validation accuracy during training
~~- [ ] possibly make it possible to pick another dataset (from torchvision for instance ?)~~
~~- [ ] command line arguments ?~~

Overall should try to stay below 300 LoC to keep this pedagogical and easy to grok

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by ClashLuke** [comment]
We should probably remove the einops dependency in the type checker, but I couldn't do that reasonably well on GitHub web.

File: examples/microViT.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                    "dim_model": self.hparams.dim,
+                    "layer_norm_style": "pre",
+                    "multi_head_config": {
+                        "num_heads": self.hparams.n_head,
+                        "residual_dropout": self.hparams.resid_pdrop,
+                        "use_rotary_embeddings": True,
+                        "attention": {
+                            "name": self.hparams.attention,
+                            "dropout": self.hparams.attn_pdrop,
+                            "causal": True,
```
Comment (SeanNaren) [inline examples/microViT.py:77]: For causal to be enabled we need to pass a seq_len right? I wasn't sure what to set the seq len (from what I gather, we need the infered size after the conv layer)

File: examples/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+pytorch-lightning-bolts
```
Comment (SeanNaren) [inline examples/requirements.txt:1]:
```suggestion
lightning-bolts
```

We need to kill the other repo, was before my time unsure why it changed! this is the repo it should be I think

File: README.md [hunk 3]
```diff
@@ -186,7 +186,7 @@ If you use xFormers in your publication, please cite it by using the following B
 
 ``` bibtex
 @Misc{xFormers2021,
-  author =       {Benjamin Lefaudeux, Francisco Massa, Diana Liskovich, Wenhan Xiong, Vittorio Caggiano, Min Xu, Jieru Hu, Marta Tintore, Susan Zhang},
+  author =       {Benjamin Lefaudeux, Francisco Massa, Diana Liskovich, Wenhan Xiong, Vittorio Caggiano, Sean Naren, Min Xu, Jieru Hu, Marta Tintore, Susan Zhang},
```
Comment (SeanNaren) [inline README.md:189]: Proud to be in this list of amazing people :)

File: examples/microViT.py (hunk 1 above)
Comment (blefaudeux) [reply examples/microViT.py:77]: ah it depends on the attentions, some of them can do without a seq len (which is really a 'max seq len'). In that case it should just be the num_patches + class token I think, I can check that out

**Comment by blefaudeux** [comment]
an afterthought is that it's probably better not to handle many datasets here, we're trying to have a minimal and simple example. Same for the command line arguments, I think that it's simple enough to be forked and altered

**Review by SeanNaren — APPROVED** [review]
Amazing :)

> an afterthought is that it's probably better not to handle many datasets here, we're trying to have a minimal and simple example. Same for the command line arguments, I think that it's simple enough to be forked and altered

Agreed! let's keep this micro

**Comment by Vittorio-Caggiano** [comment]
This is fantastic! Thanks @SeanNaren for contributing those examples. I wonder if it might be good to add a readme file also in the `/examples` folder to summarize both examples.

**Comment by blefaudeux** [comment]
> This is fantastic! Thanks @SeanNaren for contributing those examples. I wonder if it might be good to add a readme file also in the `/examples` folder to summarize both examples.

agree, will do ! I think we can also add the ViT patching from TImm using sparse attention from @fmassa here, it's pretty powerful and I could match it with cifar or mnist fashion so that it's completely runnable out of the box

**Review by Vittorio-Caggiano — APPROVED** [review]
great! LGTM!
